homeassistant/components/mqtt/client.py

```diff
@@ -706,15 +706,10 @@ async def _async_resubscribe(self) -> None:
         """Resubscribe on reconnect."""
         # Group subscriptions to only re-subscribe once for each topic.
         keyfunc = attrgetter("topic")
-        await self._async_perform_subscriptions(
-            [
-                # Re-subscribe with the highest requested qos
-                (topic, max(subscription.qos for subscription in subs))
-                for topic, subs in groupby(
-                    sorted(self.subscriptions, key=keyfunc), keyfunc
-                )
-            ]
-        )
+        self._pending_subscriptions = {
```

Comment:
There is a race here. `self.connected` is set before this runs so its possible for another `async_subscribe` call to run before and have `self._pending_subscriptions` overwritten.

Reply:
```
        self.connected = True
        dispatcher_send(self.hass, MQTT_CONNECTED)
        _LOGGER.info(
            "Connected to MQTT server %s:%s (%s)",
            self.conf[CONF_BROKER],
            self.conf[CONF_PORT],
            result_code,
        )

.....

        if (
            CONF_BIRTH_MESSAGE in self.conf
            and ATTR_TOPIC in self.conf[CONF_BIRTH_MESSAGE]
        ):

            async def publish_birth_message(birth_message: PublishMessage) -> None:
                await self._ha_started.wait()  # Wait for Home Assistant to start
                await self._discovery_cooldown()  # Wait for MQTT discovery to cool down
                await self.async_publish(
                    topic=birth_message.topic,
                    payload=birth_message.payload,
                    qos=birth_message.qos,
                    retain=birth_message.retain,
                )

            birth_message = PublishMessage(**self.conf[CONF_BIRTH_MESSAGE])
            asyncio.run_coroutine_threadsafe(
                publish_birth_message(birth_message), self.hass.loop
            )

```

I'd move this block into `_async_resubscribe` and rename it to `_async_connected`

Than there is no way that `self.connected` can be set before its run.

You'll have to change `dispatcher_send` to `async_dispatcher_send` as well dropping the `run_coroutine_threadsafe`